### ISSUE TITLE
Rewrite hierarchical gibbs sampling code to use new framework

### ISSUE THREAD

**Issue by MichaelClerx** [pints-team/pints#38]

**Comment by MichaelClerx** [comment]
Need to think how to set up the interface!

- [ ] Create example code using the (non-existent) hierarchical mcmc methods
- [ ] Write the actual implementation

**Comment by MichaelClerx** [comment]
See @martinjrobins original code!

**Comment by martinjrobins** [comment]
btw. My original code needed the following:
- have a mapping between the hyperparameters and the lower level mcmc parameters
- be able to modify the prior of the lower level mcmc log-likelihood function

**Comment by MichaelClerx** [comment]
@martinjrobbins, could we split this into two interfaces? So that you define a single problem (noise) and then wrap the meta problem around it (variability)?

**Comment by MichaelClerx** [comment]
And the first problem would just be the classes we have now

**Comment by MichaelClerx** [comment]
And the second would allow an infinite hierarchy?

**Comment by MichaelClerx** [comment]
Not sure how much this overlaps with #39 btw

**Comment by martinjrobins** [comment]
I think that a nice way to do the hierarchical sampling would be to allow priors to have hyper-parameters. For example, a normal distribution could have mean and covariance hyperparameters. If a prior did not have any hyper-parameters, then it is just treated like the prior class is at the moment. How does this sound? I would like to alter the Prior class like so, where the dimension of the hyper-parameters is allowed to be 0 (for a normal prior)

```python
class Prior(object):
    """
    Represents a prior distribution on a vector of variables.

    Given any point `x` in the parameter space, a prior function needs to be
    able to evaluate the probability density `f(x)` (such that the integral
    over `f(x)dx` equals 1).
    """
    def dimension(self):
        """
        Returns the dimension of the space this prior is defined on.
        """
        raise NotImplementedError

    def hyper_dimension(self):
        """
        Returns the dimension of the space the hyperparameters are defined on.
        """
        raise NotImplementedError

    def evaluate(self, x, hyper_params):
        """
        Returns the probability density for point `x`,
        given a set of hyper-parameters
        """
        raise NotImplementedError

    def generate(self, hyper_params):
        """
        Returns a sample point `x` from the probability density,
        """
        raise NotImplementedError
```

**Comment by MichaelClerx** [comment]
Yeah that sounds like it makes sense!
Could we find a way to keep the __call__ method though? Or make it easier for priors _not_ to have hyper params if they don't want too?

**Comment by martinjrobins** [comment]
After discussions with @sanmitraghosh a few months ago, I came to the conclusion that implementing hyperparameters in pints would be too complicated to support in general, and would only duplicate efforts in other probabilistic programming languages.

I still need to implement this hierarchical inference code, but will do it as a one-off example. Not sure how to do this in pints nicely though. It would require me using n=10 MCMC samplers in parallel (one for each of the 10 experiments), but altering the priors before each consecutive sample is taken. Is there a way to do this nicely?

**Comment by MichaelClerx** [comment]
Not yet! But that would probably be possible if we had an "ask-and-tell" interface (see #132 and rant in #185)

**Comment by martinjrobins** [comment]
I've rewritten the hierarchical mcmc to use the ask-and-tell interface, which works really well (thanks @MichaelClerx!). You can see the result here, using the logistic model:

https://github.com/martinjrobins/hierarchical_mcmc/blob/master/main.py

Here are the resultant plots:

[hchains.pdf](https://github.com/pints-team/pints/files/1784674/hchains.pdf)
[htrace.pdf](https://github.com/pints-team/pints/files/1784675/htrace.pdf)
[hpairwise.pdf](https://github.com/pints-team/pints/files/1784677/hpairwise.pdf)

The parameters for the logistic equation work really well, but the noise parameter isn't working (mean is under half what it should be). I'll have to consult @sanmitraghosh or @ben18785 on this next time we meet!

**Comment by MichaelClerx** [comment]
Looks really cool!

Do you think we could set up a generic routine for this at some point?

**Comment by martinjrobins** [comment]
I don't think so, @sanmitraghosh and I had a chat about it a while back and decided it was too difficult to implement this in general, because it depends on the particular set of conjugate priors chosen. I think this is the problem that probabilistic programming is designed to solve, and we don't want to go implementing yet another probabilistic programming language in pints. I think the ask-and-tell interface is a good compromise, it still allows you to construct hierarchical models without implementing the sampler algorithms.

**Comment by martinjrobins** [comment]
ahhhh, found a really dumb bug that was the cause of the noise parameter issues, see updated code (@chonlei). New plots below, everything looks perfect now.

Might be worth writing this up as a pints example? It is a nice case study on how useful the new ask-and-tell interface is to construct more complicated statistical models. Our answer to probabilistic programming...

[hchains.pdf](https://github.com/pints-team/pints/files/1789421/hchains.pdf)
[htrace.pdf](https://github.com/pints-team/pints/files/1789422/htrace.pdf)
[hpairwise.pdf](https://github.com/pints-team/pints/files/1789423/hpairwise.pdf)

**Comment by MichaelClerx** [comment]
Great! Would definitely turn that into an example yes!

**Comment by martinjrobins** [comment]
works well on real data too :)

[hchains.pdf](https://github.com/pints-team/pints/files/1807454/hchains.pdf)
[htrace.pdf](https://github.com/pints-team/pints/files/1807455/htrace.pdf)
[hpairwise.pdf](https://github.com/pints-team/pints/files/1807456/hpairwise.pdf)

**Comment by MichaelClerx** [comment]
Great stuff. Closing this issue now!
